packages/labs/ssr-react/src/lib/node/render-shadow-contents.ts

```diff
@@ -0,0 +1,82 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {LitElementRenderer} from '@lit-labs/ssr/lib/lit-element-renderer.js';
+import {getElementRenderer} from '@lit-labs/ssr/lib/element-renderer.js';
+import type {RenderInfo} from '@lit-labs/ssr';
+
+const reservedReactProperties = new Set([
+  'children',
+  'localName',
+  'ref',
+  'style',
+  'className',
+]);
+
+const attributesToProps = (attrs: NamedNodeMap) => {
+  const props: {[index: string]: string} = {};
+  for (let i = 0; i < attrs.length; i++) {
+    const attr = attrs[i];
+    props[attr.name] = attr.value;
+  }
+  return props;
+};
+
+/**
+ * Renders the shadow contents of the provided custom element type with props.
+ * Should only be called in server environments.
+ */
+export const renderShadowContents = (type: string, props: {} | null) => {
```

Comment:
There's _almost_ nothing React-specific about this, right? This is generally useful because it wraps up the minimum steps needed to transform a tagName, attrs, and props from a call site into the reflected attrs and SR contents using ElementRenderer (something others have stumbled over as well). e.g. the Astro integration could use the same logic.

Can we move this to `labs/ssr`, with an additional `attrs` argument, and then this file would only add specific logic for separating React props into CE attrs & props?

Reply:
Naming nit: since this is more than just rendering "shadow contents" (technically it handles props -> reflected attributes as well), `renderCustomElement` might be better? Also `type` -> `tagName` would be clearer.

Reply:
Renamed. I agree that this could very well be part of labs/ssr but want to punt it to make sure we can separately discuss the proper API for making it more general use.

Reply:
Filed https://github.com/lit/lit/issues/3656